### PR TITLE
Fix a crash while replaying `rr record rr replay`

### PR DESCRIPTION
**Pull request by Keno**
See commit message for details. First commit is just a small refactor to factor out common code used by the test in the second commit.

### CODE REVIEW

File: src/Task.cc [hunk 1]
```diff
@@ -841,15 +841,20 @@ TrapReasons Task::compute_trap_reasons() {
...
-                 .local_addr,
-             PRELOAD_THREAD_LOCALS_SIZE);
-    } else {
-      // The mapping might have been removed by crazy application code.
-      // That's OK, assuming the preload library was removed too.
-      memset(&thread_locals, 0, sizeof(thread_locals));
+      uint8_t* local_addr =
+          as->mapping_of(AddressSpace::preload_thread_locals_start())
+              .local_addr;
+      if (local_addr) {
```
Comment (rocallahan) [inline src/Task.cc]: This is a bit scary. It's possible that the mapping was replaced by something else that gets a local mapping, and then this code will run and we'll scribble over it. That would be pretty hard to debug. Can we additionally check that the name starts with `preload_thread_locals`?

Comment (Keno) [inline src/Task.cc]: I had thought about that, but I didn't think anything that would get mapped there would actually get a local mapping. However, to be safe, we could add an extra flag similar to `IS_SYSCALLBUF`.

Comment (rocallahan) [inline src/Task.cc]: > I didn't think anything that would get mapped there would actually get a local mapping

That's probably true at the moment but it's a fragile assumption.

**Comment by Keno** [comment]
Changed to use a flag.

**Comment by rocallahan** [comment]
Merged with some formatting fixes. Thanks!

**Comment by Keno** [comment]
> Merged with some formatting fixes. Thanks!

Ah yes, sorry forgot to reformat after editing.

**Pull request closed without merging**